File: 27-02-2025_Get_Min_From_Stack.py
# Python program to implement a stack that supports
# all operations in O(1) time and O(n) extra space.

class SpecialStack:
    def __init__(self):
        self.s = []

    # Add an element to the top of Stack
    def push(self, x):
        newMin = x if not self.s else min(x, self.s[-1][1])
        self.s.append((x, newMin))

    # Remove the top element from the Stack
    def pop(self):
        if self.s:
            self.s.pop()

    # Returns top element of the Stack
    def peek(self):
        return -1 if not self.s else self.s[-1][0]

    # Finds minimum element of Stack
    def getMin(self):
        return -1 if not self.s else self.s[-1][1]

if __name__ == "__main__":
    ss = SpecialStack()
    # Function calls
    ss.push(2)
    ss.push(3)
    print(ss.peek(), end=" ")
    ss.pop()
    print(ss.getMin(), end=" ")
    ss.push(1)
    print(ss.getMin(), end=" ")

File: 27-02-2025_Get_Min_From_Stack.java
// Java program to implement a stack that supports
// all operations in O(1) time and O(n) extra space.

import java.util.Stack;

class SpecialStack {
    private Stack<int[]> s;

    public SpecialStack() {
        s = new Stack<>();
    }

    // Add an element to the top of Stack
    public void push(int x) {
        int newMin = s.isEmpty() ? x : Math.min(x, s.peek()[1]);
        s.push(new int[]{x, newMin});
    }

    // Remove the top element from the Stack
    public void pop() {
        if (!s.isEmpty()) {
            s.pop();
        }
    }

    // Returns top element of the Stack
    public int peek() {
        return s.isEmpty() ? -1 : s.peek()[0];
    }

    // Finds minimum element of Stack
    public int getMin() {
        return s.isEmpty() ? -1 : s.peek()[1];
    }

    public static void main(String[] args) {
        SpecialStack ss = new SpecialStack();

        // Function calls
        ss.push(2);
        ss.push(3);
        System.out.print(ss.peek() + " ");
        ss.pop();
        System.out.print(ss.getMin() + " ");
        ss.push(1);
        System.out.print(ss.getMin() + " ");
    }
}